File: solutions/python/proverb/1/proverb.py
from typing import Optional


def proverb(*input_data: str, qualifier: Optional[str]) -> list[str]:
    qualifier_str = qualifier + " " if qualifier is not None else ""
    return [
        f"For want of a {a} the {b} was lost."
        for a, b in zip(input_data, input_data[1:])
    ] + [
        f"And all for the want of a {qualifier_str}{input_datum}."
        for input_datum in input_data[:1]
    ]

File: solutions/go/proverb/1/proverb.go
package proverb

func Proverb(rhyme []string) []string {
	if len(rhyme) == 0 {
		return []string{}
	}

	var result []string
	for i := 0; i < len(rhyme)-1; i++ {
		line := "For want of a " + rhyme[i] + " the " + rhyme[i+1] +
			" was lost."
		result = append(result, line)
	}
	finalLine := "And all for the want of a " + rhyme[0] + "."
	result = append(result, finalLine)

	return result
}
